src/cursor/cur_index.c

```diff
@@ -347,6 +354,43 @@ __curindex_search_near(WT_CURSOR *cursor, int *exact)
     API_END_RET(session, ret);
 }
 
+/*
+ * __increment_bound_array --
+ *     Increment the given buffer by one bit. If the all of the values inside the buffer is maxed
+ *     values
+ */
+static inline int
+__increment_bound_array(WT_ITEM *user_item)
+{
+    size_t usz;
+    uint8_t *userp;
+    int i;
+
+    usz = user_item->size;
+    userp = (uint8_t *)user_item->data;
+
+    if (usz <= 0)
+        return (0);
+
+    /*
+     * First loop through all max values on the buffer from the end. This is to find the appropriate
+     * position to increment add one to the byte.
+     */
+    for (i = usz - 1; i >= 0 && userp[i] == UINT8_MAX; --i)
+        ;
+
+    /*
+     * If all of the buffer are max values, we don't need to do anything as the key format is a
+     * fixed length format. Ideally we double check that the table format has a fixed length string.
+     */
+    if (i < 0)
+        return (0);
+    else
+        userp[i] += 1;
+
+    return (0);
```

Comment:
I think I miscommunicated the algorithm.  You need to increment the last byte, and if it goes to 0, you need to "carry" to the previous byte.  You had this correct in the previous version.  Where it's different from the previous version, is that if it overflows the final byte, then: 1) if it's the upper bound we're dealing with, you should remove the upper bound entirely.  That's because the only way it can overflow all the bytes is if it is a fixed size format.  2) if it's the lower bound and we're incrementing because we it's exclusive, then there is nothing within the bounds.  That's because (since it overflows), it must be fixed size, and the caller has specified a lower/exclusive bound that is the highest possible value.  It's legal I think, but there can be nothing in the range.

The key format does not have to be fixed length.  But the only way all the bytes can overflow is if it is fixed length.  You could assert that if you want.

If we want to meet, I'd be happy to do that.  I could also draft the algorithm if that helps.

Reply:
Let me know if this is the algorithm you thought of!